Make pg_dump 17.6 compatible with ActiveRecord

eanders opened the pull request:
## _Merging this PR_
- use the squash-merge strategy for PRs targeting a release-X branch

## Description
[//]: # (Summarize changes and include links related issue)
[//]: # (List any new dependencies or relevant ADRs)
* Fixes an issue introduced with pg_dump 17.6 by commenting out lines in the structure dump files that start with \restrict or \unrestrict

## Type of change
[//]: # (e.g., Bug fix, New feature, Documentation, Code clean-up, Dependency update)
* Bug fix

## Checklist before requesting review
[//]: # (Remove any items that are not applicable)
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail

eanders committed:
Make pg_dump 17.7 compatible with ActiveRecord

eanders committed:
Don't replace the entire line for \unrestrict

eanders committed:
Centralize pg_dump monkey patching

eanders committed:
Don't try to fix non-existent files

ttoomey reviewed (approved):
Pretty hacky but probably the best we can do without sinking more time. Too bad we can't just pin postgres to 17.4 locally

It looks like there's a ton of churn in the sql files, can you roll that back or rebuild from a fresh test db to reduce the churn?

Also surprised that `schema.gsub!(/^\\restrict/, '-- \restrict')` doesn't need to be a mutliline (`\m`) but I guess it works

eanders committed:
Rebuild dbs

eanders commented:
Rebuilt with:
```
╰─$ git checkout release-180 -- db/structure.sql
╭─elliot@host ~/Sites/op/hmis-warehouse ‹ruby-3.3.8› ‹ea-postgres-17-6-fix●› 
╰─$ git checkout release-180 -- db/warehouse_structure.sql
╭─elliot@host ~/Sites/op/hmis-warehouse ‹ruby-3.3.8› ‹ea-postgres-17-6-fix●› 
╰─$ git checkout release-180 -- db/reporting_structure.sql
╭─elliot@host ~/Sites/op/hmis-warehouse ‹ruby-3.3.8› ‹ea-postgres-17-6-fix●› 
╰─$ git checkout release-180 -- db/health_structure.sql   
╭─elliot@host ~/Sites/op/hmis-warehouse ‹ruby-3.3.8› ‹ea-postgres-17-6-fix●› 
╰─$ dcr spec                                           
[+] Creating 3/3
 ✔ Container hmis-warehouse-redis  Running                                                                                                                                        0.0s 
 ✔ Container hmis-warehouse-db     Running                                                                                                                                        0.0s 
 ✔ Container s3                    Running                                                                                                                                        0.0s 
app-user@cd3d24814463:/app$ rails db:drop:primary
Dropped database 'warehouse_app_test'
app-user@cd3d24814463:/app$ rails db:drop:warehouse
Dropped database 'warehouse_test'
app-user@cd3d24814463:/app$ rails db:drop:health
Dropped database 'health_test'
app-user@cd3d24814463:/app$ rails db:drop:reporting
Dropped database 'reporting_test'
app-user@cd3d24814463:/app$ rails db:create:primary && rails db:create:warehouse && rails db:create:health && rails db:create:reporting
Created database 'warehouse_app_test'
Created database 'warehouse_test'
Created database 'health_test'
Created database 'reporting_test'
app-user@cd3d24814463:/app$ bundle exec rails db:structure:conditional_load:primary
app-user@cd3d24814463:/app$ bundle exec rails db:migrate:primary
app-user@cd3d24814463:/app$ bundle exec rails db:structure:conditional_load:reporting 
app-user@cd3d24814463:/app$ bundle exec rails db:structure:conditional_load:health
app-user@cd3d24814463:/app$ bundle exec rails db:structure:conditional_load:warehouse
app-user@cd3d24814463:/app$ bundle exec rails db:migrate:health && bundle exec rails db:migrate:warehouse && bundle exec rails db:migrate:reporting
```
I guess we have some discrepancies in 180 as well, but hopefully this is closer.
